drivers/hmis/lib/hmis_errors.rb

```diff
@@ -0,0 +1,57 @@
+###
+# Copyright 2016 - 2022 Green River Data Analysis, LLC
+#
+# License detail: https://github.com/greenriver/hmis-warehouse/blob/production/LICENSE.md
+###
+
+module HmisErrors
+  class CustomValidationError
+    def initialize(attribute, type = :invalid, message: nil, full_message: nil, severity: :error, readable_attribute: nil, **kwargs)
+      {
+        attribute: attribute,
+        type: type,
+        message: message,
+        full_message: full_message,
+        readable_attribute: readable_attribute,
+        severity: severity,
+        **kwargs,
+      }.each do |key, value|
+        define_singleton_method(key) { value }
+      end
+    end
+
+    def to_h
+      {
+        attribute: attribute,
+        type: type,
+        message: message,
+        full_message: full_message,
+        severity: severity,
+        readable_attribute: readable_attribute,
+      }
+    end
+
+    def warning?
+      severity&.to_sym == :warning
+    end
+  end
+
+  class CustomValidationErrors
+    include Enumerable
+
+    extend Forwardable
+    def_delegators :@errors, :size, :clear, :blank?, :empty?, :uniq!, :any?, :count
+    attr_reader :errors
+    alias objects errors
+
+    def initialize
+      @errors = []
+    end
+
+    def add(attribute, type = :invalid, **options)
+      error = CustomValidationError.new(attribute, type, **options)
+      @errors.append(error)
+      error
+    end
```

Comment:
maybe?
```suggestion
    def add(attribute, type = :invalid, **options)
      error = CustomValidationError.new(attribute, type, **options)
      @errors.append(error)
      error
    end
    alias :<< :add
```

Reply:
How would that be used? I get a syntax error with this: `errors << :field_name, :invalid`

Reply:
Not sure it's worth figuring out, but I liked the previous syntax, something like:
`errors << CustomValidationError.new(:field_name, :invalid, **options)`

Reply:
I prefer `.add` because it is shorter. These error lines get so long that the condition wraps, which is annoying to read but apparently unavoidable.